packages/pyright-internal/src/common/uri/uri.ts

```diff
@@ -0,0 +1,206 @@
+/*
+ * uri.ts
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ *
+ * URI namespace for storing and manipulating URIs.
+ */
+
+import { URI, Utils } from 'vscode-uri';
+import { EmptyUri } from './emptyUri';
+import { FileUri } from './fileUri';
+//import * as memoization from './memoization_instrumenting';
+import { WebUri } from './webUri';
+
+export interface Uri {
+    // Unique key for storing in maps.
+    readonly key: string;
+
+    // Returns the scheme of the URI.
+    readonly scheme: string;
+
+    // Returns the last segment of the URI, similar to the UNIX basename command.
+    readonly basename: string;
+
+    // Returns the extension of the URI, similar to the UNIX extname command.
+    readonly extname: string;
+
+    // Returns a URI where the path just contains the root folder.
+    readonly root: Uri;
+
+    // Returns a URI where the path contains the directory name with .py appended.
+    readonly packageUri: Uri;
+
+    // Returns a URI where the path contains the directory name with .pyi appended.
+    readonly packageStubUri: Uri;
+
+    // Returns a URI where the path has __init__.py appended.
+    readonly initFileUri: Uri;
+
+    // Returns a URI where the path has __init__.pyi appended.
+    readonly initStubUri: Uri;
+
+    // Returns a URI where the path has py.typed appended.
+    readonly pytypedUri: Uri;
+
+    // Returns the basename without any extensions
+    readonly withoutExtension: string;
+
+    isEmpty(): boolean;
+
+    toString(): string;
+
+    toUserVisibleString(): string;
+
+    matchesRegex(regex: RegExp): boolean;
+
+    replaceExtension(ext: string): Uri;
+
+    addExtension(ext: string): Uri;
+
+    addPath(extra: string): Uri;
+
+    // Returns a URI where the path is the directory name of the original URI, similar to the UNIX dirname command.
+    getDirectory(): Uri;
+
+    getRootPathLength(): number;
+
+    // Determines whether a path consists only of a path root.
+    isRoot(): boolean;
+
+    // Determines whether a Uri is a child of some parent Uri.
+    isChild(parent: Uri, ignoreCase?: boolean): boolean;
+
+    isLocal(): boolean;
+
+    isUntitled(): boolean;
+
+    equals(other: Uri | undefined, ignoreCase?: boolean): boolean;
+
+    startsWith(other: Uri | undefined, ignoreCase?: boolean): boolean;
+
+    pathStartsWith(name: string): boolean;
+
+    pathEndsWith(name: string): boolean;
+
+    pathIncludes(include: string): boolean;
+
+    // How long the path for this Uri is.
+    getPathLength(): number;
+
+    combinePaths(...paths: string[]): Uri;
+
+    getRelativePath(child: Uri): string | undefined;
+
+    getPathComponents(): readonly string[];
+
+    getPath(): string;
+
+    getFilePath(): string;
+
+    getRelativePathComponents(to: Uri): readonly string[];
+    getShortenedFileName(maxDirLength?: number): string;
+
+    stripExtension(): Uri;
```

Comment:
Can we organize these methods into related groups? For example, all of the `fooExtension` methods should be together.